### PR TITLE
cpu/atxmega/periph: Add i2c driver

### PR DESCRIPTION
**Pull request by nandojve**
### Contribution description

Add ATxmega i2c master driver. The atxmega-a1u-xpro and atxmega-a3bu-xplained boards received default EXT-1/J1 header config.

### Testing procedure

The test was conducted using tests/periph_i2c using a at24mac device.

### Issues/PRs references

This is part of #15703

CC: @benpicco @maribu

### CODE REVIEW

File: cpu/atxmega/periph/i2c.c [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+/**
+ * @brief   Device Context allocation
+ */
+static struct {
+    mutex_t locks;
+    sema_t xfer;
```
Comment (benpicco) [inline cpu/atxmega/periph/i2c.c:38]: Why a semaphore and not just a mutex?

File: cpu/atxmega/periph/i2c.c [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    uint8_t baudrate;
+
+    if (i2c >= I2C_NUMOF) {
+        DEBUG("[i2c] init: dev is invalid.\n");
+        return;
+    }
+
+    baudrate = _i2c_calc_baud(i2c);
+    if (baudrate == 0) {
+            return;
```
Comment (benpicco) [inline cpu/atxmega/periph/i2c.c:85]: maybe add an `assert(0)` here

File: cpu/atxmega/periph/i2c.c (hunk 1 above)
Comment (nandojve) [reply cpu/atxmega/periph/i2c.c]: Reworked to use mutex.

File: cpu/atxmega/periph/i2c.c (hunk 2 above)
Comment (nandojve) [reply cpu/atxmega/periph/i2c.c]: I miss a debug message.

File: cpu/atxmega/Makefile.dep [hunk 3]
```diff
@@ -4,6 +4,10 @@ USEMODULE += atxmega_periph
 # All ATxmega based CPUs provide PM
 USEMODULE += pm_layered
 
+ifeq (,$(filter periph_i2c,$(USEMODULE)))
+USEMODULE += posix_semaphore
```
Comment (benpicco) [inline cpu/atxmega/Makefile.dep:8]: no longer needed

File: cpu/atxmega/periph/i2c.c [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+ * @brief   Device Context allocation
+ */
+static struct {
+    mutex_t locks;
+    mutex_t xfer;
+    uint8_t *buffer;
+    size_t pos;
+    size_t len;
+    uint8_t flags;
+    volatile int8_t status;
```
Comment (benpicco) [inline cpu/atxmega/periph/i2c.c:42]: No need for `volatile` here, you are waiting on `xfer` before accessing it.

**Review by benpicco — COMMENTED** [review]
Looks good!
What result do you get from the `i2c_scan` command?

File: cpu/atxmega/Makefile.dep (hunk 3 above)
Comment (nandojve) [reply cpu/atxmega/Makefile.dep]: yeah...sorry!

File: cpu/atxmega/periph/i2c.c (hunk 4 above)
Comment (nandojve) [reply cpu/atxmega/periph/i2c.c]: removed!

**Comment by nandojve** [comment]
> Looks good!
> What result do you get from the `i2c_scan` command?

```
> i2c_scan 0
2021-04-23 21:47:25,340 # i2c_scan 0
2021-04-23 21:47:25,342 # Scanning I2C device 0...
2021-04-23 21:47:25,349 # addr not ack'ed = "-", addr ack'ed = "X", addr reserved = "R", error = "E"
2021-04-23 21:47:25,352 #      0 1 2 3 4 5 6 7 8 9 a b c d e f
2021-04-23 21:47:25,355 # 0x00 R R R R R R R R R R R R R R E E
2021-04-23 21:47:25,359 # 0x10 E E E E E E E E E E E E E E E E
2021-04-23 21:47:25,362 # 0x20 E E E E E E E E X E E E E E E E
2021-04-23 21:47:25,365 # 0x30 X E E E E E E E E E E E E E E E
2021-04-23 21:47:25,368 # 0x40 E E E E E E E E E E E E E E E E
2021-04-23 21:47:25,372 # 0x50 X E E E E E E E X E E E E E E E
2021-04-23 21:47:25,375 # 0x60 E E E E E E E E E E E E E E E E
2021-04-23 21:47:25,378 # 0x70 E E E E E E E E R R R R R R R R
> i2c_acquire 0
2021-04-23 21:47:30,969 # i2c_acquire 0
2021-04-23 21:47:30,971 # Command: i2c_acquire(0)
2021-04-23 21:47:30,973 # Success: i2c_0 acquired
i2c_write_byte 0 0x58 0x98 0
2021-04-23 21:47:38,024 # i2c_write_byte 0 0x58 0x98 0
2021-04-23 21:47:38,028 # Command: i2c_write_byte(0, 0x58, 0x00, [0x98])
2021-04-23 21:47:38,031 # Success: i2c_0 wrote 1 byte to the bus
i2c_read_bytes 0 0x58 16 0
2021-04-23 21:47:42,952 # i2c_read_bytes 0 0x58 16 0
2021-04-23 21:47:42,956 # Command: i2c_read_bytes(0, 0x58, 16, 0x00)
2021-04-23 21:47:42,968 # Success: i2c_0 read 16 byte(s) : [0xfc, 0xc2, 0x3d, 0x00, 0x00, 0x0d, 0xdc, 0xf1, 0x0a, 0x90, 0x85, 0x04, 0x94, 0x10, 0x04, 0x90]
> i2c_release 0
2021-04-23 21:47:49,774 # i2c_release 0
2021-04-23 21:47:49,776 # Command: i2c_release(0)
2021-04-23 21:47:49,778 # Success: i2c_0 released
>
```

**Comment by benpicco** [comment]
Something seems to be wrong, you get an error for an address that has not been ack'ed.
Should look more like this

```
2021-04-24 15:19:19,561 # Scanning I2C device 0...
2021-04-24 15:19:19,567 # addr not ack'ed = "-", addr ack'ed = "X", addr reserved = "R", error = "E"
2021-04-24 15:19:19,571 #      0 1 2 3 4 5 6 7 8 9 a b c d e f
2021-04-24 15:19:19,574 # 0x00 R R R R R R R R R R R R R R - -
2021-04-24 15:19:19,579 # 0x10 - - - - - - - - - - - - - - - -
2021-04-24 15:19:19,584 # 0x20 - - - - - - - - - - - - - - - -
2021-04-24 15:19:19,589 # 0x30 X - - - - - - - - - - - - - - -
2021-04-24 15:19:19,594 # 0x40 - - - - - - - - - - - - - - - -
2021-04-24 15:19:19,599 # 0x50 X - - - - - - - X - - - - - - -
2021-04-24 15:19:19,604 # 0x60 - - - - - - - - - - - - - - - -
2021-04-24 15:19:19,608 # 0x70 - - - - - - - - R R R R R R R R
```

**Comment by benpicco** [comment]
Looks good, please squash!

**Comment by nandojve** [comment]
rebase and squash.

**Review by benpicco — APPROVED** [review]
The latest fixup looks like it will produce the correct output for `i2c_scan`, so ACK

**edit** looks like two `Makefile.ci`s now need an update

**Pull request merged**